fix(socket): Fix default socket directory for MacOS.

Klafyvel opened the pull request:
Fix #14. We need someone using MacOS to confirm that it works. :)

mihrits commented:
I am the author of the screenshot in #14. I tried this PR, by switching the branch in lazy.nvim. It seems to have updated to the right commit. My Julia package is added as `REPLSmuggler.jl#main`. Unfortunately my neovim still doesn't detect any sockets to connect to. `netstat` does show open socket.

Klafyvel commented:
Thanks your update! Can you give me the output of
`BaseDirs.User.runtime(REPLSmuggler.PROJECT)`?

On Tue, Aug 20, 2024, 13:34 Moorits Mihkel Muru ***@***.***>
wrote:

> I am the author of the screenshot in #14
> <https://github.com/Klafyvel/nvim-smuggler/issues/14>. I tried this PR,
> by switching the branch in lazy.nvim. It seems to have updated to the right
> commit. My Julia package is added as REPLSmuggler.jl#main. Unfortunately
> my neovim still doesn't detect any sockets to connect to. netstat does
> show open socket.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Klafyvel/nvim-smuggler/pull/23#issuecomment-2298637327>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABPL4L63B5NYKR6WWHUYJXDZSMSTXAVCNFSM6AAAAABMKTN25GVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDEOJYGYZTOMZSG4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

mihrits commented:
I included the `versioninfo()` as well, just in case.
```
julia> versioninfo()
Julia Version 1.10.4
Commit 48d4fd48430 (2024-06-04 10:41 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: macOS (arm64-apple-darwin22.4.0)
  CPU: 8 × Apple M1 Pro
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-15.0.7 (ORCJIT, apple-m1)
Threads: 1 default, 0 interactive, 1 GC (on 6 virtual cores)
Environment:
  JULIA_EDITOR = nvim

julia> BaseDirs.User.runtime(REPLSmuggler.PROJECT)
"/Users/moorits/Library/Application Support/lang.julia.REPLSmuggler/"
```

Klafyvel commented:
It may very well be, can you try adding it in the vim plugin?

mihrits commented:
I am not sure I did it correctly. Anyway, I found the plugin source code on the disk. Added the slash. Lazy.nvim complains that it can't sync, because there are local modifications. So I guess I made the modification in the correct file. But I am not sure if lazy.nvim loads the change files or just complains. In any case, it seems that it didn't make a difference. I can open the SmuggleConfig menu, but it is still empty.

I am not even sure, if that was the correct way to do it. Should I have forked the branch, made the change and added that as a new plugin?

Klafyvel commented:
My understanding of things is that should have been enough. Can you check
that REPLSmuggler does create the file at the correct location?

On Tue, Aug 20, 2024, 15:07 Moorits Mihkel Muru ***@***.***>
wrote:

> I am not sure I did it correctly. Anyway, I found the plugin source code
> on the disk. Added the slash. Lazy.nvim complains that it can't sync,
> because there are local modifications. So I guess I made the modification
> in the correct file. But I am not sure if lazy.nvim loads the change files
> or just complains. In any case, it seems that it didn't make a difference.
> I can open the SmuggleConfig menu, but it is still empty.
>
> I am not even sure, if that was the correct way to do it. Should I have
> forked the branch, made the change and added that as a new plugin?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Klafyvel/nvim-smuggler/pull/23#issuecomment-2298817712>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABPL4L5PJUG4EXQBL5Y7NVLZSM5P5AVCNFSM6AAAAABMKTN25GVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDEOJYHAYTONZRGI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

mihrits commented:
I also found that `vim.fn.has("unix") returns `1` for me. A google search suggested that `has("unix")` might not work to differentiate between Linux and MacOS. I tried changing the `unix` to `linux`, that didn't change anything. But switching the `has("mac")` to first if statement made the difference that now I do see a socket in the menu. However, it seems I still can't send code to the REPL. Every time I try to send something I get a prompt to choose a socket and nothing happens after that.

Klafyvel commented:
Ok! we're making progress. I forced pushed the fix as it should currently be on your computer. We can try to debug a bit further by enabling debug logs. Could you pull the new changes, and then switch the boolean `debug_enabled` in `lua/smuggler/config.lua` line 7 to `true` and then restart neovim? nvim-smuggler should create a smuggler.log file where it dumps some information on what it is doing.

mihrits commented:
`vim.fn.has("linux")` returns 0 or 1, but it turns out that Lua handles both of these as truths in boolean context. Adding `vim.fn.has("linux") == 1` helped get the socket to the list without switching the order. Unfortunately, still unable to forward the code to REPL.

Klafyvel commented:
ah, damn. Lua really has some weird features.

mihrits commented:
It does create a `smuggler.log` file, but it is empty. Even though I tried sending the code to REPL.

Klafyvel commented:
Does it remain empty once you quit neovim?

mihrits commented:
Yes, it does remain empty even after quitting.

Klafyvel commented:
I see, could you change line 114 of `lua/smuggler/protocol.lua` to 
```lua
  socket:connect(socket_path, function(err)
    if err ~= nil then
      error("Error while connecting:" .. err)
    end
  end)
```
I think the connection fails for some reason.

mihrits commented:
I made the change in `protocol.lua`, but no changes in the outcome. The log file is still empty. No errors showing in neovim either.

Klafyvel commented:
I'm running out of ideas :D can you change the `error` call to a `config.debug` call? You could also have a look at what the buffer configuration looks like: `:lua vim.print(require("smuggler.config").buf)`.

mihrits commented:
Hmm, `:lua vim.print(require("smuggler.config").buf)` returned `{}`.

Changing the `error` to `config.debug` didn't do anything.

Just in case, I'll copy my config for loading the plugin with lazy.nvim as well.
```lua
return {
  "klafyvel/nvim-smuggler",
  branch = "fix-macos-default-socket-path",
  config = true,
  opts = {
    mappings = true, -- set to false to disable all mappings.
    map_smuggle = "<leader>js", -- Use `[count]<leader>js` in normal mode to send count lines.
    map_smuggle_range = "<leader>js", -- Use `<leader>js` in visual mode to send the current selection.
    map_smuggle_config = "<leader>je", -- Use `<leader>je` in normal mode to reconfigure the plugin.
    map_smuggle_operator = "gjs", -- Use `gjs[text object]` to send a text object in normal mode.
  },
  dependencies = { "nvim-neotest/nvim-nio" },
}
```

Klafyvel commented:
> Hmm, `:lua vim.print(require("smuggler.config").buf)` returned `{}`.

Ok! If that was after having chosen a socket, this means we never reach the end of the `runclient` call in protocol.lua! Which means something crashes before that. You could try to insert various `config.debug` calls in that function to see what is the last line that gets executed (perhaps we don't even enter the function if `choosesocket` fails).

mihrits commented:
I added a `config.debug` to `chooseclient` and now the debugger has
```
"Chose socket: "
nil
```
So I guess I see the socket listed, but can't actually choose it? Maybe I am choosing it wrong? I tried opening another REPLSmuggler socket. I see them both listed, I can switch between the options, but choosing one (<CR>) seems to just close the menu.

Klafyvel commented:
I see. The socket choice is a bit broken anyway, I'll work on it to fix #24 once I'm done with #15 (should be a few days). In the mean time, I think you could hard-code the paths in REPLSmuggler.jl and nvim-smuggler if you want to try out the plugin. (replace the function call with a fixed path in the neovim plugin and call REPLSmuggler with [this method call](https://github.com/Klafyvel/REPLSmuggler.jl/blob/main/src/REPLSmuggler.jl#L102) : `smuggle("my_scoket", basepath="~")` for example. :)

ArbitRandomUser commented:
main is not the branch you should be trying out for this ?

@mihrits you mentioned you added #main , but you should be using the branch fix-macos-default-socket-path

ArbitRandomUser commented:
it looks like an issue from #24 becuase from the screenshot it does look like @mihrits is using `telescope-ui-select`

Klafyvel commented:
> it looks like an issue from #24 becuase from the screenshot it does look like @mihrits is using `telescope-ui-select`

Oh, indeed, I failed to notice that. `telescope-ui-select` does not work (yet) because it is asynchronous. I'll fix it soon.

Klafyvel commented:
Hi! I rebased this PR to include my recent work on refactoring the module and fixing `telescope-ui-select` integration. I hope this now works for you.

mihrits commented:
Thanks! I can confirm that everything seems to work out of the box now. I was able to connect to the socket and smuggle code to the REPL.
